Add the Swan Lake Update 4 blog (#152)
ballerina-platform/blog: _posts/2023-02-16-announcing-ballerina-2201.4.0-swan-lake-update-4.md

---
layout: blog-post
title: Announcing Ballerina 2201.4.0 (Swan Lake Update 4)
author: Ballerina Team

> published-date: 16 December 2023

Reviewer: Date should be corrected right

Author: Fixed. Thanks.

status: Published
socialmediaimage: Ballerina-Swan-Lake-GA-Release-banner-02-with-button.png
permalink: /posts/2023-02-16-announcing-ballerina-2201.4.0-swan-lake-update-4/
---

<style>.cBlogContent p{white-space: break-spaces !important;}</style>

> We are happy to announce a new [Ballerina 2201.4.0 (Swan Lake Update 4)]((https://ballerina.io/downloads/),) release, which is the fourth update release of Ballerina Swan Lake.

Reviewer 2: ```suggestion
We are happy to announce a new [Ballerina 2201.4.0 (Swan Lake Update 4)](https://ballerina.io/downloads/) release, which is the fourth update release of Ballerina Swan Lake.
```

> The main highlight of this release is to support creating multiple output bindings when using Azure Functions. From this release onwards, you can define a tuple as a return type and annotate each member of the tuple using an output binding. This requirement is facilitated from the compiler level by the introduction of the ability for tuple members to be annotated via `field` attachment point declarations.

Reviewer 3: Let's move this to a separate section. I guess we don't have a key highlight in this release

Author: Will be addressed via https://github.com/ballerina-platform/blog/pull/152/commits/e192e27c931f57913c8b99809bc566cb57e1fc88.

> The main highlight of this release is to support creating multiple output bindings when using Azure Functions. From this release onwards, you can define a tuple as a return type and annotate each member of the tuple using an output binding. This requirement is facilitated from the compiler level by the introduction of the ability for tuple members to be annotated via `field` attachment point declarations.

Reviewer 2: ```suggestion
The main highlight of this release is support for creating multiple output bindings when using Azure Functions. From this release onward, you can define a tuple as a return type and annotate each member of the tuple using an output binding. This requirement is facilitated by the language by the introduction of support for tuple members to be annotated using annotations declared with the `field` attachment point.
```

Reviewer 2: Don't we usually add examples in the release blog?

Author: IMHO, the blog should be a summary of the main highlights of the release and examples should be added in the release note or in Learn pages, BBEs, or API Docs.

The rest of the significant additions and improvements are described below.

## Runtime updates

> Now, you can obtain the strand dump when running the tests in a Ballerina package or a file using the bal test command by sending the `SIGTRAP` signal to that process. The `FiniteType` runtime class is updated to return a user-defined type name if available when the `getName()` method is called on it. Also, runtime APIs are now modified to return the type-reference type instances according to the type definitions.

Reviewer 2: ```suggestion
Now, you can obtain the strand dump when running the tests in a Ballerina package or a file using the `bal test` command and sending the `SIGTRAP` signal to that process. The `FiniteType` runtime class is updated to return a user-defined type name if available when the `getName()` method is called on it. Also, runtime APIs are now modified to return the type-reference type instances according to the type definitions.
```

> Now, you can obtain the strand dump when running the tests in a Ballerina package or a file using the `bal test` command and sending the `SIGTRAP` signal to that process. The `FiniteType` runtime class is updated to return a user-defined type name if available when the `getName()` method is called on it. Also, runtime APIs are now modified to return the type-reference type instances according to the type definitions.

Reviewer 4: ```suggestion
Now, you can obtain the strand dump when running the tests in a Ballerina package or a file using the `bal test` command and sending the `SIGTRAP` signal to that process. 
```

Author: Will be addressed via https://github.com/ballerina-platform/blog/pull/152/commits/e192e27c931f57913c8b99809bc566cb57e1fc88.

## Standard library updates

> - The Update 4 release introduces the `ballerina/persist` package, which provides the Ballerina persistence layer functionality of conveniently storing and querying data in a data store. New persist CLI tooling is also introduced to enable the Ballerina persistence layer in a Ballerina project to generate types and clients based on the data model. This is an experimental feature. The APIs and tooling experience might change in future releases.

Reviewer 3: ```suggestion
- The Update 4 release introduces the experimental `ballerina/persist` package, providing a mechanism to manage persistent data conveniently. Describe the entities, attributes, and relationships in your data model using familiar type definition syntax in Ballerina. Then, generate the client API to query and manipulate data using the new utilities available in the `bal` tool. 
```

Author: Will be addressed via https://github.com/ballerina-platform/blog/pull/152/commits/e192e27c931f57913c8b99809bc566cb57e1fc88.

> - The Update 4 release introduces the `persist package`, which provides the Ballerina persistent layer functionality of conveniently storing and querying data in a data store. New persist CLI tooling is also introduced along with this to enable the Ballerina persistence layer in a Ballerina project to generate types and clients based on the data model. This is an experimental feature. The APIs and tooling experience might change in future releases.

Reviewer 2: I thought we weren't mentioning persistence in this release? 

```suggestion
- The Update 4 release introduces the `ballerina/persist` package, which provides the Ballerina persistence layer functionality of conveniently storing and querying data in a data store. New persist CLI tooling is also introduced to enable the Ballerina persistence layer in a Ballerina project to generate types and clients based on the data model. This is an experimental feature. The APIs and tooling experience might change in future releases.
```

- The `graphql` package is improved such that now, a single client can have multiple subscriptions (multiplexing), and if additional information about a particular field of a GraphQL document (e.g., subfields of a field) is required, that information can be accessed inside a resolver or an interceptor method.
- A new JetStream client is introduced to the `nats` package to carry out the publishing and subscribing streaming functionalities related to the NATS JetStreams feature.
- The `constraint` package is enhanced by adding constraint validation support for read-only types.
- The `oauth2` package now allows the use of inferred values for refresh token configurations through the password grant type, and string values for the `scopes` field in the client grant configuration.

## Code to cloud updates

Now, you can disable creating thin JAR files when generating the Docker image in order to reduce the build time of the generation. However, this will disable the Docker layering feature.  Also, the `CMD` property of the Dockerfile is improved to support complex package names.

## Developer tools updates

> - This Update 4 release introduces a new `Create variable with Type` code action along with `Quick Pick` support for selecting expressions in the `Extract to constant` code action. Additionally, the `Rename popup` feature was added to the `Extract to constant`, `Extract to local variable`, and 'Extract to function' code actions. 

Reviewer 2: ```suggestion
- This Update 4 release introduces a new `Create variable with Type` code action along with `Quick Pick` support for selecting expressions in the `Extract to constant` code action. Additionally, the `Rename popup` feature was added to the `Extract to constant`, `Extract to local variable`, and `Extract to function` code actions. 
```

> - From this release onwards, you can generate Ballerina client and service declarations from Swagger 2.0 (i.e., OpenAPI 2.0) definitions along with a few other improvements done for OpenAPI object schema attributes and the default request and response types of the generated Ballerina resource/remote methods.

Reviewer 5: Shall we split into two sentences as below, as having a single sentence might make the idea complicated IMO?
```suggestion
- From this release onwards, you can generate Ballerina client and service declarations from Swagger 2.0 (i.e., OpenAPI 2.0) definitions. Along with that, also a few other improvements were introduced which are related to OpenAPI object schema attributes and, the default request and response types of the generated Ballerina resource/remote methods.
```

Author: Updated.

> - New options are introduced to the `bal format` command to format modules or single files selectively with a provided Ballerina package. Also, the `--native` flag is introduced to the `bal test` command for executing tests using a GraalVM native executable image. This is an experimental feature.

Reviewer 3: ```suggestion
- New options are introduced to the `bal format` command to format modules or single files selectively with a provided Ballerina package. 
- The `--native` flag is introduced to `bal test` command for executing tests using a GraalVM native executable image. This is an experimental feature.
```

Author: Will be addressed via https://github.com/ballerina-platform/blog/pull/152/commits/e192e27c931f57913c8b99809bc566cb57e1fc88.

- The GraphQL tool can now generate GraphQL SDL schema files from GraphQL services for users to share those files for documentation or any other purpose as required.

## Ballerina packages updates

> From this Update 4 release onwards, now, you can maintain code generated by other tools in a Ballerina package. This feature allows you to keep the generated source code in a separate directory named ‘generated’ within the package structure while treating those files as part of the Ballerina package source.

Reviewer 2: Shall we recheck if we really need "now" wherever it has been used? Redundant here since we start with "From this Update 4 release onwards"?

Actually even that sentence has both this and update 4, one is kinda redundant?

```suggestion
From this release onward, you can maintain code generated by other tools in a Ballerina package. This feature allows you to keep the generated source code in a separate directory named ‘generated’ within the package structure while treating those files as part of the Ballerina package source.
```

Author: Updated.

> For a detailed list of what’s new in Swan Lake Update 4, see the [release note](https://ballerina.io/downloads/swan-lake-release-notes/swan-lake-2201.4.0).

Reviewer 2: ```suggestion
For a detailed list of what's new in Swan Lake Update 4, see the [release note](https://ballerina.io/downloads/swan-lake-release-notes/swan-lake-2201.4.0).
```
